src/replication.c

```diff
@@ -3216,15 +3222,42 @@ void setupMainConnForPsync(connection *conn) {
                                      "accept connections in read-write mode.\n");
         }
         dualChannelSyncHandlePsync();
-        return;
+        return C_OK;
     }
 
-error:
+    return psync_result;
+}
+
+/* Replication: Replica side.
+ * This connection handler fires after rdb-connection was initialized. We use it
+ * to adjust the replica main for loading incremental changes into the local buffer. */
+void dualChannelSetupMainConnForPsync(connection *conn) {
+    char *err = NULL;
+    int ret;
+
+    switch (server.repl_state) {
+    case REPL_STATE_SEND_HANDSHAKE: ret = dualChannelReplMainConnSendHandshake(conn, &err); break;
```

Comment:
This is what I meant by consolidating state manipulations in the "driver" function and then all the "worker" functions (such as `dualChannelReplMainConnRecvCapaReply`) become "stateless".
```suggestion
    case REPL_STATE_SEND_HANDSHAKE: 
        ret = dualChannelReplMainConnSendHandshake(conn, &err); 
        if (ret != C_ERR) server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
        /* else server.repl_state = XXX; // if it makes sense */
        break;
```

Reply:
sounds good